File: 4949.cpp
#include "iostream"
#include "string"

int main()
{
    std::string ans;
    bool flag = true;
    while(flag){
        while(true){
            char stack[100];
            int len = 0;
            int c;
            scanf("%c",&c);
            if(c == '\0'){
                flag = false;
                break;
            }
            if(c == '.'){
                ans += "yes";
                break;
            }
            else if(c == '('){
                stack[len] = '(';
                len++;
            }
            else if(c == '['){
                stack[len] = '[';
                len++;
            }
            else if(c == ')'){
                if(stack[len] != '('){
                    ans += "no";
                    break;
                }
            }
            else if(c == ']'){
                if(stack[len] != '['){
                    ans += "no";
                    break;
                }
            }
        }
    }
    
    std::cout << ans;    
}

File: 4949.py
ans = ""

while True:
    stack = list(0 for i in range(100))
    len = 1
    sentence = input()
    if sentence == ".":
        break
    for c in sentence:
        if c == '.':
            ans += "yes\n"
            break
        elif c == '[':
            stack += "["
            len += 1
        elif c == '(':
            stack += "("
            len += 1
        elif c == ']':
            if stack[len-1] != '[':
                ans += "no\n"
                break
            else:
                len -= 1
        elif c == ')':
            if stack[len-1] != '(':
                ans += "no\n"
                break
            else:
                len -= 1

print(ans)
        